### DISCUSSION TITLE
Nested queries: making an atomic write dependent on the result of a read

### DISCUSSION THREAD

**Discussion by jamesonev** [prisma/prisma#5694] [Q&A]
Hi there,

I'm having a hard time figuring out how to combine atomic operations with nested reads. 

I'm trying to create a new tuple based on the value of a findFirst query plus some constant, but I need a transactional guarantee. I can do it sequentially (not transactionally) like this:

```ts
const readQuery = await prisma.balanceTuple.findFirst({
  where: {
          balance: { not: null },
          user: { id: myUserId },
        },
        select: {
          balance: true,
        },
        orderBy: { createdAt: 'desc' }
 });

const writeQuery = await prisma.balanceTuple.create({
  data: {
         amount: 100,
          balance: readQuery.balance  + 100,
          user: {connect: { id: myUserId }}
        }
 });
```
But if my balance changes between the read and write, that data will be lost.
I want to use a nested query, but I can't figure out how to specify my read query. This is my best shot so far:
```ts
 prisma.balanceTuple.create({
    data: {
      amount: 100,
      balance: {
        100 +
        findFirst: {
            where: {
              balance: { not: null },
              user: { id: myUserId },
            },
            select: {
              balance: true,
            },
            orderBy: { createdAt: 'desc' },
          },
      },
     user: {connect: { id: myUserId }}
    },
  });
```

Thanks!

**Comment by ryands17** [comment]
@jamesonev 👋 

You can use [atomic number operations](https://www.prisma.io/docs/reference/api-reference/prisma-client-reference#increment-all-view-and-likes-fields-of-all-post-records-by-1) for this. So your example would look like:

```ts
const writeQuery = await prisma.balanceTuple.create({
  data: {
         amount: 100,
          balance: { increment: 100 },
          user: {connect: { id: myUserId }}
        }
 });
```

**Comment by jamesonev** [comment]
Thanks for the suggestion, I looked into atomic operations but I'm trying to create a new tuple _without_ modifying the original record and that's where I'm stuck.

**Reply by ryands17** [reply]
In this case, I would suggest using [Optimistic Concurrency Control](https://www.prisma.io/docs/guides/prisma-guides/prisma-client-transactions-guide#optimistic-concurrency-control) to only fetch the value and only update if the value matches the one that you fetched to maintain integrity.

**Comment by ludralph** [comment]
Hi there,

To keep our discussions organized and focused on the most relevant topics, we’re reviewing and tidying up our backlog. As part of this process, we’re closing discussions that haven’t had any recent activity and appear to be outdated.

If this discussion is still important to you or unresolved, we’d love to hear from you! Feel free to reopen it or start a new one with updated details.

For more details about our priorities and vision for the future of Prisma ORM, check out our latest blog post: https://www.prisma.io/blog/prisma-orm-manifesto.

Thank you for your understanding and being part of the community!
